pkg/reconciler/trustedresources/verify.go

```diff
@@ -50,3 +60,138 @@ func verifyInterface(
+	if err != nil {
+		return err
+	}
+	pipeline := v1beta1.Pipeline{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: "tekton.dev/v1beta1",
+			Kind:       "Pipeline"},
+		ObjectMeta: pm,
+		Spec:       pipelineObj.PipelineSpec(),
+	}
+	verifiers, err := getVerifiers(ctx)
+	if err != nil {
+		return err
+	}
+
+	for _, verifier := range verifiers {
+		if err := VerifyInterface(ctx, pipeline, verifier, signature); err == nil {
+			return nil
+		}
+	}
+
+	return fmt.Errorf("Pipeline %v in namespace %v fails verification", pipeline.Name, pipeline.Namespace)
+}
+
+// prepareObjectMeta will filter the objectmeta and extract the signature.
+func prepareObjectMeta(in metav1.ObjectMeta) (metav1.ObjectMeta, []byte, error) {
+	out := metav1.ObjectMeta{}
+
+	// exclude the fields populated by system.
+	out.Name = in.Name
+	out.GenerateName = in.GenerateName
+	out.Namespace = in.Namespace
+
+	if in.Labels != nil {
+		out.Labels = make(map[string]string)
+		for k, v := range in.Labels {
+			out.Labels[k] = v
+		}
+	}
+
+	out.Annotations = make(map[string]string)
+	for k, v := range in.Annotations {
+		out.Annotations[k] = v
+	}
+
+	// exclude the annotations added by other components
+	// Task annotations are unlikely to be changed, we need to make sure other components
+	// like resolver doesn't modify the annotations, otherwise the verification will fail
+	delete(out.Annotations, "kubectl-client-side-apply")
+	delete(out.Annotations, "kubectl.kubernetes.io/last-applied-configuration")
+
+	// signature should be contained in annotation
+	sig, ok := in.Annotations[SignatureAnnotation]
+	if !ok {
+		return out, nil, fmt.Errorf("signature is missing")
+	}
+	// extract signature
+	signature, err := base64.StdEncoding.DecodeString(sig)
+	if err != nil {
+		return out, nil, err
+	}
+	delete(out.Annotations, SignatureAnnotation)
+
+	return out, signature, nil
+}
+
+// getVerifiers get all verifiers from configmap
+func getVerifiers(ctx context.Context) ([]signature.Verifier, error) {
+	cfg := config.FromContextOrDefaults(ctx)
+	verifiers := []signature.Verifier{}
+
+	for key := range cfg.TrustedResources.Keys {
+		v, err := loadPublicKey(ctx, key)
+		if err == nil {
+			verifiers = append(verifiers, v)
+		}
+	}
+
+	if len(verifiers) == 0 {
+		return verifiers, fmt.Errorf("no public keys are founded for verification")
+	}
+
+	return verifiers, nil
+}
+
+// loadPublicKey is a wrapper for VerifierForKeyRef, hardcoding SHA256 as the hash algorithm
+func loadPublicKey(ctx context.Context, keyRef string) (verifier signature.Verifier, err error) {
+	return verifierForKeyRef(ctx, keyRef, crypto.SHA256)
+}
+
+// verifierForKeyRef parses the given keyRef, loads the key and returns an appropriate
+// verifier using the provided hash algorithm
+func verifierForKeyRef(ctx context.Context, keyRef string, hashAlgorithm crypto.Hash) (verifier signature.Verifier, err error) {
+	raw, err := os.ReadFile(filepath.Clean(keyRef))
```

Comment:
Might just be me being paranoid, but you may want to pull this up and guarantee that the same verifiers are used for the same reconcile event. Otherwise you may get race conditions if the key changes mid-reconcile (though I suspect this will be rare).

I'm okay if you want to hold off on this / TODO for now though!

Reply:
I think this is a great suggestion!
Which level should I pull this up? Moving this to `reconcile` and pass verifiers down to the place where we will use them?

Reply:
Maybe it you can wrap the resolvers with a verifier when they are created? 🤔